src/js/components/FileInput/FileInput.js

```diff
@@ -221,198 +295,146 @@ const FileInput = forwardRef(
+                            inputRef.current.click();
+                            inputRef.current.focus();
+                          }}
+                          label={format({
+                            id: 'fileInput.browse',
+                            messages,
+                          })}
+                          disabled={files.length >= max}
+                        />
+                      )}
+                    </Keyboard>
+                  )}
+                </Box>
               </Box>
-            </Box>
-          ))}
-        <StyledFileInput
-          ref={inputRef}
-          type="file"
-          id={id}
-          name={name}
-          multiple={multiple}
-          disabled={disabled}
-          plain
-          rightOffset={rightOffset}
-          {...rest}
-          onDragOver={() => setDragOver(true)}
-          onDragLeave={() => setDragOver(false)}
-          onChange={(event) => {
-            event.persist();
-            const fileList = event.target.files;
-            const nextFiles = multiple ? [...files] : [];
-            for (let i = 0; i < fileList.length; i += 1) {
-              // avoid duplicates
-              const existing =
-                nextFiles.filter(
-                  (file) =>
-                    file.name === fileList[i].name &&
-                    file.size === fileList[i].size,
-                ).length > 0;
-              if (!existing) nextFiles.push(fileList[i]);
-            }
-            setFiles(nextFiles);
-            setDragOver(false);
-            if (onChange) onChange(event, { files: nextFiles });
-          }}
-        />
-      </ContentsBox>
+            ))}
+          <StyledFileInput
+            ref={inputRef}
+            type="file"
+            id={id}
+            name={name}
+            multiple={multiple}
+            disabled={disabled}
+            plain
+            rightOffset={rightOffset}
+            {...rest}
+            onDragOver={() => setDragOver(true)}
+            onDragLeave={() => setDragOver(false)}
+            onChange={(event) => {
+              event.persist();
+              const fileList = event.target.files;
+              let nextFiles = multiple ? [...files] : [];
+              for (let i = 0; i < fileList.length; i += 1) {
+                // avoid duplicates
+                const existing =
+                  nextFiles.filter(
+                    (file) =>
+                      file.name === fileList[i].name &&
+                      file.size === fileList[i].size,
+                  ).length > 0;
+                if (!existing) nextFiles.push(fileList[i]);
+              }
+              if (nextFiles.length > max) {
+                nextFiles = nextFiles.slice(0, max);
+              }
+              setFiles(nextFiles);
+              setDragOver(false);
+              if (onChange) onChange(event, { files: nextFiles });
+            }}
+          />
+        </ContentsBox>
+        {multiple?.max && (
```

Comment:
This should be integrated with FormField `validate` and `help` rather than adding the message directly here.

Take a look at how https://github.com/grommet/grommet/pull/5457 is handling a `maxSize` for FileInput.

Reply:
I have made the changes but I am not sure about the following things:-
-  how are we gonna display the `maxFile` message when the `FileInput` is not wrapped inside the `FormField` i.e in other stories.
- User will still saw the validate error message even when we remove `required` prop from the FormField. I know this is because of this `if (fileList?.length && fileList.length <= max) return;`. But is it possible if I can know if the user has added `required` prop or not ?

If you see any other issue, please let me know about it.

Reply:
I'm not sure of the answer to this yet. The `required` prop and validations from the component don't really play well together. Typically both should get executed in order and the first one that fails wins. But this doesn't seem to be how the form does it.

Also, given that the browse button even let you add more than the limit, I don't see the validator or message ever being useful. It will never show except this odd empty field case.

Reply:
Then, what do you guys think is the good approach to tackle this ?

Reply:
After talking to Eric, what we want to do is remove the code that truncates the list of files that comes from the Browse dialog. Just let the user add however many files they choose then let the validation show that is over the limit.

The the other change is to use remove the `useFormField` call and replace it with a new version of useFormInput that can provide the validation function and error text. This will have to wait for #5457 which adds this ability. This will make `required` still work correctly plus any other validation added to the `FormField`

Reply:
> After talking to Eric, what we want to do is remove the code that truncates the list of files that comes from the Browse dialog. Just let the user add however many files they choose then let the validation show that is over the limit.
> 
> Then the other change is to use remove the `useFormField` call and replace it with a new version of useFormInput that can provide the validation function and error text. This will have to wait for #5457 which adds this ability. This will make `required` still work correctly plus any other validation added to the `FormField`

This code is still disabling browse based on `files.length >= max`. The decision was to let users pick a bunch of files and just rely on the validation to mark it invalid. If you disable the browse then the user wouldn't be able to correct it.